File: nim/oop.nim
type
  Animal = ref object of RootObj
  Dog = ref object of Animal
  Cat = ref object of Animal

method hello(a: Animal): string {.base.} =
  discard

method hello(dog: Dog): string =
  return "Bark"

method hello(cat: Cat): string =
  return "Meow"

var animals = [Dog(), Cat()]
for a in animals:
  echo a.hello()

File: rust/oop.rs
trait Animal {
    fn hello(&self) -> String;
}

struct Dog;
impl Animal for Dog {
    fn hello(&self) -> String {
        String::from("Bark")
    }
}

struct Cat;
impl Animal for Cat {
    fn hello(&self) -> String {
        String::from("Meow")
    }
}

fn main() {
    let animals = [&Dog as &Animal, &Cat as &Animal];
    for a in &animals {
        println!("{}", a.hello());
    }
}
